cgen error on init fixed size array of generic type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 2a69b7c, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 f048bb3.2a69b7c
|:-------------------|:-------------------
|OS                  |linux, "EndeavourOS Linux"
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 8845HS w/ Radeon 780M Graphics
|Memory              |19.6GB/30.64GB
|                    |
|V executable        |/home/lmp/Projects/v/v
|V last modified time|2025-01-21 08:39:50
|                    |
|V home dir          |OK, value: /home/lmp/Projects/v
|VMODULES            |OK, value: /home/lmp/.vmodules
|VTMP                |OK, value: /dev/shm
|Current working dir |OK, value: /home/lmp/Projects/v
|                    |
|Git version         |git version 2.48.1
|V git status        |weekly.2025.03-19-g2a69b7c5 (2 commit(s) behind V master)
|.git/config present |true
|                    |
|cc version          |cc (GCC) 14.2.1 20240910
|gcc version         |gcc (GCC) 14.2.1 20240910
|clang version       |clang version 19.1.7
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.72-git (437140d149d9c977ffc8b09dbaf9b0f5a02db190)
|glibc version       |ldd (GNU libc) 2.40

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  t.v && t`
```v
module main

import math.vec

pub struct Bezier {
pub mut:
	points [4]vec.Vec2[f32]
}

pub fn (b Bezier) h3() vec.Vec2[f32] {
	return b.points[2]
}

pub fn (b Bezier) h4() vec.Vec2[f32] {
	return b.points[3]
}

pub fn (b Bezier) end() vec.Vec2[f32] {
	return b.h4()
}

pub struct Path {
pub mut:
	segments [1024]Bezier
	len      int // should be read-only
}

pub fn (mut p Path) add(b Bezier) {
	assert p.len + 1 < 1024
	p.segments[p.len] = b
	p.len++
}

fn main() {
	b1 := Bezier{
		points: [vec.Vec2[f32]{100, 100}, vec.Vec2[f32]{120, 80},
			vec.Vec2[f32]{200, 70}, vec.Vec2[f32]{400, 200}]!
	}
	b2 := Bezier{
		points: [b1.end(), b1.h3() + vec.Vec2[f32]{0, 2 * 80}, vec.Vec2[f32]{200, 70},
			b1.end() + vec.Vec2[f32]{400, 200}]!
	}

	mut path := Path{}

	path.add(b1)
	path.add(b2)

	println(sizeof(path))
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /dev/shm/t.01JJ4PA9HFFVSK3Y4YDY0MWBMP.tmp.c:7745: error: cannot convert 'struct math__vec__Vec2_T_f32' to 'void *'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Something like "32772"

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2769,9 +2769,12 @@ fn (mut g Gen) expr_with_fixed_array(expr ast.Expr, got_type_raw ast.Type, expec
 		// [ foo(), foo() ]!
 		val_typ := g.table.value_type(got_type_raw)
 		val_styp := g.styp(val_typ)
+		val_sym := g.table.final_sym(val_typ)
+		prefix := if val_sym.kind !in [.array, .array_fixed] { '&' } else { '' }
 		for i, item_expr in expr.exprs {
-			g.write('memcpy(&${tmp_var}[${i}], ')
-			needs_addr := (item_expr is ast.CallExpr && !item_expr.return_type.is_ptr())
+			g.write('memcpy(${prefix}${tmp_var}[${i}], ')
+			needs_addr := (item_expr is ast.CallExpr && !item_expr.return_type.is_ptr()
+				&& g.table.final_sym(item_expr.return_type).kind !in [.array, .array_fixed])
 				|| (item_expr is ast.InfixExpr && !item_expr.promoted_type.is_ptr())
 				|| item_expr is ast.StructInit
 			if needs_addr {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2768,10 +2768,20 @@ fn (mut g Gen) expr_with_fixed_array(expr ast.Expr, got_type_raw ast.Type, expec
 		g.writeln('${styp} ${tmp_var};')
 		// [ foo(), foo() ]!
 		val_typ := g.table.value_type(got_type_raw)
+		val_styp := g.styp(val_typ)
 		for i, item_expr in expr.exprs {
-			g.write('memcpy(${tmp_var}[${i}], ')
+			g.write('memcpy(&${tmp_var}[${i}], ')
+			needs_addr := (item_expr is ast.CallExpr && !item_expr.return_type.is_ptr())
+				|| (item_expr is ast.InfixExpr && !item_expr.promoted_type.is_ptr())
+				|| item_expr is ast.StructInit
+			if needs_addr {
+				g.write('ADDR(${val_styp}, ')
+			}
 			g.expr(item_expr)
-			g.writeln(', sizeof(${g.styp(val_typ)}));')
+			if needs_addr {
+				g.write(')')
+			}
+			g.writeln(', sizeof(${val_styp}));')
 		}
 	} else if expr is ast.CallExpr {
 		// return var.call() where returns is option/result fixed array

--- a/vlib/v/tests/builtin_arrays/array_fixed_init_node_test.v
+++ b/vlib/v/tests/builtin_arrays/array_fixed_init_node_test.v
@@ -1,0 +1,50 @@
+module main
+
+import math.vec
+
+pub struct Bezier {
+pub mut:
+	points [4]vec.Vec2[f32]
+}
+
+pub fn (b Bezier) h3() vec.Vec2[f32] {
+	return b.points[2]
+}
+
+pub fn (b Bezier) h4() vec.Vec2[f32] {
+	return b.points[3]
+}
+
+pub fn (b Bezier) end() vec.Vec2[f32] {
+	return b.h4()
+}
+
+pub struct Path {
+pub mut:
+	segments [1024]Bezier
+	len      int // should be read-only
+}
+
+pub fn (mut p Path) add(b Bezier) {
+	assert p.len + 1 < 1024
+	p.segments[p.len] = b
+	p.len++
+}
+
+fn test_main() {
+	b1 := Bezier{
+		points: [vec.Vec2[f32]{100, 100}, vec.Vec2[f32]{120, 80},
+			vec.Vec2[f32]{200, 70}, vec.Vec2[f32]{400, 200}]!
+	}
+	b2 := Bezier{
+		points: [b1.end(), b1.h3() + vec.Vec2[f32]{0, 2 * 80}, vec.Vec2[f32]{200, 70},
+			b1.end() + vec.Vec2[f32]{400, 200}]!
+	}
+
+	mut path := Path{}
+
+	path.add(b1)
+	path.add(b2)
+
+	assert sizeof(path) == 32772
+}

